File: 프로그래머스/lv2/최원준/게임 맵 최단거리.py
# https://www.acmicpc.net/problem/

'''
1. 아이디어 :
    bfs로 풀 수 있다
2. 시간복잡도 :
    O( n * m )
3. 자료구조 :
    배열
'''


from collections import deque
def solution(maps):
    visited = set()
    visited.add((0,0))
    queue = deque()
    queue.append((0,0,1)) #row, col, dist
    dir = [[0,-1],[0,1],[-1,0],[1,0]]
    while queue:
        r1, c1, dist = queue.popleft()
        if r1 == len(maps)-1 and c1 == len(maps[0])-1:
            return dist

        for r2, c2 in dir:
            nr, nc = r1+r2, c1+c2
            if not(0<=nr<len(maps) and 0<=nc<len(maps[0])) or (nr,nc) in visited or maps[nr][nc] == 0:
                continue
            visited.add((nr,nc))
            queue.append((nr,nc,dist+1))

    return -1

File: 프로그래머스/lv2/지준혁/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    int[] dy = {-1, 0, 1, 0};
    int[] dx = {0, 1, 0, -1};
    public int solution(int[][] maps) {
        
        return bfs(maps);
    }
    
    int bfs(int[][] maps) {
        
        int n = maps.length;
        int m = maps[0].length;
        int[][] dist = new int[n][m];
        for (int i = 0; i < n; i++) Arrays.fill(dist[i], -1);
        Queue<int[]> q = new LinkedList<>();
    
        dist[0][0] = 1;
        q.offer(new int[]{0, 0});
        
        while (!q.isEmpty()) {
            var cur = q.poll();
            for (int dir = 0; dir < 4; ++dir) {
                int ny = cur[0] + dy[dir];
                int nx = cur[1] + dx[dir];
                
                if (ny < 0 || ny >= n || nx < 0 || nx >= m) continue;
                if (maps[ny][nx] == 0 || dist[ny][nx] != -1) continue;
                dist[ny][nx] = dist[cur[0]][cur[1]] + 1;
                q.offer(new int[]{ny, nx});
            }
        }

        return dist[n - 1][m - 1];
    }
}
